File: src/ecu_simulation/BatteryModule/src/GenerateFrames.cpp
#include "../include/GenerateFrames.h"
#include "../include/BatteryModuleLogger.h"

GenerateFrames::GenerateFrames(int socket)
{
    this->addSocket(socket);
}

int GenerateFrames::getSocket()
{
    return this->socket;
}

struct can_frame GenerateFrames::createFrame(int &id,  std::vector<uint8_t> &data, FrameType frameType)
{
    struct can_frame frame;
    switch (frameType)
    {
        case DATA_FRAME:
            frame.can_id = id & CAN_EFF_MASK;
            frame.can_dlc = data.size();
            for (uint8_t byte = 0; byte < frame.can_dlc; byte++)
            {
                frame.data[byte] = data[byte];
            }
            break;
        case REMOTE_FRAME:
            frame.can_id = id & CAN_EFF_MASK;
            frame.can_id |= CAN_RTR_FLAG;
            frame.can_dlc = data.size();
            for (uint8_t byte = 0; byte < frame.can_dlc; byte++)
            {
                frame.data[byte] = data[byte];
            }
            break;
        default:
            LOG_ERROR(batteryModuleLogger.GET_LOGGER(), "Invalid frame type");
            throw std::invalid_argument("Invalid frame type");
    }
    return frame;
}

bool GenerateFrames::sendFrame(int id, std::vector<uint8_t > data, FrameType frameType)
{
    struct can_frame frame = createFrame(id, data, frameType);
    int nbytes = write(this->socket, &frame, sizeof(frame));
    if (nbytes != sizeof(frame))
    {
        /* std::cout<<"Write error\n"; */
        LOG_WARN(batteryModuleLogger.GET_LOGGER(), "Write error in sendFrame\n");
        return -1;
    }
    return 0;
}

void GenerateFrames::addSocket(int socket)
{
    if (socket >= 0)
    {
        this->socket = socket;
        return;
    }
    /* std::cout<<"Error: Pass a valid Socket\n"; */
    LOG_WARN(batteryModuleLogger.GET_LOGGER(), "Error: Pass a valid Socket\n");
    exit(EXIT_FAILURE);
}

void GenerateFrames::sessionControl(int id, uint8_t sub_function, bool response)
{
    std::vector<uint8_t> data(3);
    if (!response)
    {
        data = {0x2,0x10,sub_function};
        this->sendFrame(id,data);
        return;
    }
    data = {0x2,0x50,sub_function};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::ecuReset(int id, bool response)
{
    std::vector<uint8_t> data(3);
    if (!response)
    {
        data = {0x2,0x11,0x3};
        this->sendFrame(id, data);
        return;
    }
    data = {0x2,0x51,0x3};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::authenticationRequestSeed(int id, const std::vector<uint8_t> &seed)
{
    if (seed.size() == 0)
    {
        std::vector<uint8_t> data = {0x03, 0x29, 0x1};
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {(uint8_t)(seed.size() + 2), 0x69, 0x1};
    for (uint8_t data_seed: seed)
    {
        data.push_back(data_seed);
    }
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::authenticationSendKey(int id, const std::vector<uint8_t> &key)
{
    if (key.size() > 0 )
    {
        std::vector<uint8_t> data = {(uint8_t)(key.size() + 2), 0x29, 0x2};
        for (uint8_t data_key: key)
        {
            data.push_back(data_key);
        }
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {0x02,0x69,0x02};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::routineControl(int id, uint8_t sub_function, uint16_t routin_identifier, bool response)
{
    if (!response)
    {
        std::vector<uint8_t> data = {0x4, 0x31, sub_function, (uint8_t)(routin_identifier / 0x100), (uint8_t)(routin_identifier % 0x100)};
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {0x04,0x71, sub_function, (uint8_t)(routin_identifier / 0x100), (uint8_t)(routin_identifier % 0x100)};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::testerPresent(int id, bool response)
{
    if (!response)
    {
        std::vector<uint8_t> data = {0x02,0x3E,0x00};
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {0x02,0x7E,0x00};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::readDataByIdentifier(int id,int identifier, std::vector<uint8_t> response )
{
    if (response.size() == 0)
    {
        std::vector<uint8_t> data = {0x03, 0x22, (uint8_t)(identifier/0x100), (uint8_t)(identifier%0x100)};
        this->sendFrame(id, data);
        return;
    }
    int length_response = response.size();
    if (length_response <= 4)
    {
        std::vector<uint8_t> data = {(uint8_t)(length_response + 3), 0x62, (uint8_t)(identifier/0x100), (uint8_t)(identifier%0x100)};
        for (uint8_t data_response: response)
        {
            data.push_back(data_response);
        }
        this->sendFrame(id, data);
        return;
    }
    /* std::cout<<"ERROR: The frame is to long!, consider using method ReadDataByIdentifierLongResponse\n"; */
    LOG_WARN(batteryModuleLogger.GET_LOGGER(), "ERROR: The frame is to long!, consider using method ReadDataByIdentifierLongResponse\n");
    return;
    /*
    *According to documentation, the frame can
    *request more than one DID. Future implementation
    *  |
    *  V
    */
}

void GenerateFrames::generateFrameLongData(int id, uint8_t sid, uint16_t identifier, std::vector<uint8_t> response, bool first_frame)
{
    if (first_frame)
    {
        std::vector<uint8_t> data = {0x10, (uint8_t)(response.size() + 3), sid, uint8_t(identifier/0x100), uint8_t(identifier%0x100)};
        /* Send only 3 first bytes of data */
        for (int byte = 0; byte < 3; byte++)
        {
            data.push_back(response[byte]);
        }
        this->sendFrame(id, data);
        return;
    }
    else
    {
        /* Delete first 3 data that were sended in the first frame */
        response.erase(response.begin(), response.begin() + 3);
        int size_of_response = response.size();
        std::vector<uint8_t> data;
        for (int no_frame = 0; no_frame <= size_of_response / 7; no_frame++)
        {
            data = {(uint8_t)(0x21 + (no_frame % 0xF))};
            for (int byte = 0; byte < 7 && ((no_frame*7) + byte) < size_of_response; byte++)
            {
                data.push_back(response[(no_frame*7)+byte]);
            }
            this->sendFrame(id, data);
        }
        return;
    } 
}

void GenerateFrames::readDataByIdentifierLongResponse(int id,uint16_t identifier, std::vector<uint8_t> response, bool first_frame)
{
    generateFrameLongData(id,0x62,identifier,response,first_frame);
}

void GenerateFrames::flowControlFrame(int id)
{
    this->sendFrame(id, {0x30,0x00,0x00,0x00});
}

void GenerateFrames::readMemoryByAddress(int id, int memory_address, int memory_size, std::vector<uint8_t> response )
{
    /* add lengths of of memory size/address to the frame */
    uint8_t length_memory_size = (countDigits(memory_size) +1) / 2;
    uint8_t length_memory_address = (countDigits(memory_address) + 1) / 2;
    uint8_t length_memory = length_memory_size * 0x10 + length_memory_address;
    if (response.size() == 0)
    {
        uint8_t pci_l = length_memory_size + length_memory_address + 2;
        std::vector<uint8_t> data = {pci_l, 0x23, length_memory};
        /* add memory address and size to the frame */
        insertBytes(data, memory_address, length_memory_address);
        insertBytes(data, memory_size, length_memory_size);
        this->sendFrame(id, data);
        return;
    }
    uint8_t pci_l = length_memory_size + length_memory_address + 2 + response.size();
    std::vector<uint8_t> data = {pci_l, 0x63, length_memory};
    /* add memory address and size to the frame */
    insertBytes(data, memory_address, length_memory_address);
    insertBytes(data, memory_size, length_memory_size);
    if (data.size() + response.size() < 9)
    {
        for (uint8_t data_response: response)
        {
            data.push_back(data_response);
        }
        this->sendFrame(id, data);
        return;
    } else
    {
        /* std::cout<<"ERROR: Response to long, consider using ReadMemoryByAdressLongResponse method\n"; */
        LOG_WARN(batteryModuleLogger.GET_LOGGER(), "ERROR: Response to long, consider using ReadMemoryByAdressLongResponse method\n");
        return;
    }
}

void GenerateFrames::readMemoryByAddressLongResponse(int id, int memory_address, int memory_size, std::vector<uint8_t> response, bool first_frame)
{
    /* add lengths of of memory size/address to the frame */
    uint8_t length_memory_size = (countDigits(memory_size) +1) / 2;
    uint8_t length_memory_address = (countDigits(memory_address) + 1) / 2;
    uint8_t length_memory = length_memory_size * 0x10 + length_memory_address;
    if (first_frame)
    {
        uint8_t pci_l = (int)response.size() + 2 + length_memory_size + length_memory_address;
        std::vector<uint8_t> data = {0x10, pci_l, 0x63, length_memory};
        /* add memory address and size to the frame */
        insertBytes(data, memory_address, length_memory_address);
        insertBytes(data, memory_size, length_memory_size);
        /* Send only 3 first bytes of data */
        for (std::size_t byte = 0; byte < (data.size() - 8); byte++)
        {
            data.push_back(response[byte]);
        }
        this->sendFrame(id, data);
        return;
    }
    else
    {
        int memory_allready_send = 8 - (length_memory_address + length_memory_size + 4);
        /* Delete data allready sended */
        response.erase(response.begin(), response.begin() + memory_allready_send);
        int size_of_response = response.size();
        std::vector<uint8_t> data;
        for (int no_frame = 0; no_frame <= size_of_response / 7; no_frame++)
        {
            data = {(uint8_t)(0x21 + (no_frame % 0xF))};
            for (int byte = 0; byte < 7 && ((no_frame*7) + byte) < size_of_response; byte++)
            {
                data.push_back(response[(no_frame*7)+byte]);
            }
            this->sendFrame(id, data);
        }
        return;
    } 
    return;
}

void GenerateFrames::writeDataByIdentifier(int id, uint16_t identifier, std::vector<uint8_t> data_parameter )
{
    if (data_parameter.size() > 0)
    {
        if (data_parameter.size() <= 4)
        {
            std::vector<uint8_t> data = {(uint8_t)(data_parameter.size() + 3),0x2E, (uint8_t)(identifier/0x100),(uint8_t)(identifier%0x100)};
            for (uint8_t data_byte: data_parameter)
            {
                data.push_back(data_byte);
            }
            this->sendFrame(id, data);
            return;
        }
        else
        {
            /* std::cout<<"The data_parameter is to long. Consider using WriteDataByIdentifierLongData method\n"; */
            LOG_WARN(batteryModuleLogger.GET_LOGGER(), "The data_parameter is to long. Consider using WriteDataByIdentifierLongData method\n");
            return;
        }
    }
    std::vector<uint8_t> data = {0x03, 0x6E, (uint8_t)(identifier/0x100),(uint8_t)(identifier%0x100)};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::writeDataByIdentifierLongData(int id, uint16_t identifier, std::vector<uint8_t> data_parameter, bool first_frame)
{
    generateFrameLongData(id,0x2E,identifier,data_parameter,first_frame);
}

void GenerateFrames::readDtcInformation(int id, uint8_t sub_function, uint8_t dtc_status_mask)
{
    std::vector<uint8_t> data = {0x03, 0x19, sub_function, dtc_status_mask};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::readDtcInformationResponse01(int id, uint8_t status_availability_mask, uint8_t dtc_format_identifier, uint8_t dtc_count)
{
    std::vector<uint8_t> data = {0x03, 0x59, 0x01, status_availability_mask, dtc_format_identifier, dtc_count};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::clearDiagnosticInformation(int id, std::vector<uint8_t> group_of_dtc, bool response)
{
    std::vector<uint8_t> data;
    /* Request */
    if (!response)
    { 
        if (group_of_dtc.size() < 7)
        {
            uint8_t number_of_dtc = group_of_dtc.size();
            data = {(uint8_t)(number_of_dtc + 1), 0x14};
            for (uint8_t dtc: group_of_dtc)
            {
                data.push_back(dtc);
            }
            this->sendFrame(id, data);
            return;
        } else
        {
            /* std::cout<<"ERROR: Can't send more than 6 DTC/frame, please consider send 2 or more frames\n"; */
            LOG_WARN(batteryModuleLogger.GET_LOGGER(), "ERROR: Can't send more than 6 DTC/frame, please consider send 2 or more frames\n");
            return;
        }
        
    }
    /* Response */
    data = {0x01, 0x54};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::accessTimingParameters(int id, uint8_t sub_function, bool response)
{
    if (!response)
    {
        std::vector<uint8_t> data = {0x02, 0x83, sub_function};
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {0x02, 0xC3, sub_function};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::negativeResponse(int id, uint8_t sid, uint8_t nrc)
{
    std::vector<uint8_t> data = {0x03, 0x7F, sid, nrc};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::requestDownload(int id, uint8_t data_format_identifier, int memory_address, int memory_size)
{
    /* Request Frame add lengths of of memory size/address to the frame */
    uint8_t length_memory_size = countDigits(memory_size + 1) / 2;
    uint8_t length_memory_address = countDigits(memory_address + 1) / 2;
    uint8_t length_memory = length_memory_size * 0x10 + length_memory_address;
    uint8_t pci_length = length_memory_size + length_memory_address + 3;
    std::vector<uint8_t> data = {pci_length, 0x34, data_format_identifier, length_memory};
    /* add memory address and size to the frame */
    insertBytes(data, memory_address, length_memory_address);
    insertBytes(data, memory_size, length_memory_size);
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::requestDownloadResponse(int id, int max_number_block)
{
    /* Response frame */
    uint8_t length_max_number_block = (countDigits(max_number_block) + 1) / 2;
    std::vector<uint8_t> data = {(uint8_t)(length_max_number_block + 2), 0x74, (uint8_t)(length_max_number_block * 0x10)};
    insertBytes(data, max_number_block, length_max_number_block);
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::transferData(int id, uint8_t block_sequence_counter, std::vector<uint8_t> transfer_request)
{
    /* If is not a response */
    if (transfer_request.size() != 0)
    {
        if (transfer_request.size() <= 5)
        {
             std::vector<uint8_t> data = {(uint8_t)(transfer_request.size() + 2), 0x36, block_sequence_counter};
            for (uint8_t data_transfer: transfer_request)
            {
                data.push_back(data_transfer);
            }
            this->sendFrame(id, data);
            return;
        } else
        {
            /* std::cout<<"The transfer_request is to long. Consider using transferDataLong method\n"; */
            LOG_WARN(batteryModuleLogger.GET_LOGGER(), "The transfer_request is to long. Consider using transferDataLong method\n");
            return;
        }
    }
    /* Response frame */
    std::vector<uint8_t> data = {0x02,0x76,block_sequence_counter};
    this->sendFrame(id, data);
    return;
}

void GenerateFrames::transferDataLong(int id, uint8_t block_sequence_counter, std::vector<uint8_t> transfer_request, bool first_frame)
{
    if (first_frame)
    {
        std::vector<uint8_t> data = {0x10, (uint8_t)(transfer_request.size() + 2), 0x36, block_sequence_counter};
        for (int byte = 0; byte < 4 ; byte++)
        {
            data.push_back(transfer_request[byte]);
        }
        this->sendFrame(id, data);
        return;
    }
    else
    {
        /* Delete first 3 data that were sended in the first frame */
        transfer_request.erase(transfer_request.begin(), transfer_request.begin() + 4);
        int size_of_data = transfer_request.size();
        std::vector<uint8_t> data;
        for (int no_frame = 0; no_frame <= size_of_data / 7; no_frame++)
        {
            data = {(uint8_t)(0x21 + (no_frame % 0xF))};
            for (int byte = 0; byte < 7 && ((no_frame*7) + byte) < size_of_data; byte++)
            {
                data.push_back(transfer_request[(no_frame*7)+byte]);
            }
            this->sendFrame(id, data);
        }
        return;
    } 
}

void GenerateFrames::requestTransferExit(int id, bool response)
{
    if (!response)
    {
        std::vector<uint8_t> data = {0x01, 0x37};
        this->sendFrame(id, data);
        return;
    }
    std::vector<uint8_t> data = {0x01,0x77};
    this->sendFrame(id, data);
    return;
}

bool GenerateFrames::requestUpdateStatus(int id, bool response)
{
    /* No impplementation, I don't find this service in the standart */
    return false;
}

/* Private */
int GenerateFrames::countDigits(int number)
{
    int digits = 0;
    if (number < 0) 
        return 0;
    while (number) {
        number /= 10;
        digits++;
    }
    return digits;
}

void GenerateFrames::insertBytes(std::vector<uint8_t>& byteVector, unsigned int num, int numBytes) {
    for (int i = numBytes - 1; i >= 0; --i) {
        byteVector.push_back((num >> (i * 8)) & 0xFF);
    }
}

File: src/rest_api/app/generate_frames.py
import can 

class GenerateFrame:
    def __init__(self, can_interface):
        self.bus = can.interface.Bus(channel=can_interface, bustype='socketcan')

    
    def send_frame(self, id, data):
        message = can.Message(arbitration_id=id, data=data)
        try: 
            self.bus.send(message)
        except can.CanError:
            print("Message not sent")

    def session_control(self, id, sub_funct, response = False):
        
        if response == False:
            data = [2, 0x10, sub_funct]
        else:
            data = [2, 0x50, sub_funct]

        self.send_frame(id, data) 
    
    def ecu_reset(self, id, response = False):

        if response == False:
            data = [2, 0x11, 0x03]
        else:
            data = [2, 0x51, 0x03]

        self.send_frame(id, data)

    def read_data_by_identifier(self, id, identifier, response = []):
        if len(response) == 0:
            
            data = [3, 0x22, identifier//0x100, identifier%0x100]
        else:
            if len(response) <= 4:

                data = [len(response) + 3, 0x62, identifier//0x100, identifier%0x100] + response

            else:
                print("Error")
                return

        self.send_frame(id, data)

    def read_data_by_identifier_long(self, id, identifier, response, first_frame=True):
        self.__generate_long_response(id, 0x62, identifier, response, first_frame)

    def request_read_dtc_information(self, id, sub_funct, dtc_status_mask):
        
        sub_funct = 0x01
       
        data = [3, 0x19, sub_funct, dtc_status_mask]

        self.send_frame(id, data)


    def response_read_dtc_information(self, id, sts_ava_mask, dtc_format, dtc_count):

        data = [3, 0x19, 0x01, sts_ava_mask, dtc_format, dtc_count]

        self.send_frame(id, data)

    def read_memory_by_adress(self, id, memory_address, memory_size, response = []):
        address_length = (self.__count_digits(memory_address)+1)//2
        size_length = (self.__count_digits(memory_size)+1)//2

        address_size_length = address_length + size_length*0x10
        pci_l = address_length + size_length + 2


        
        if len(response) == 0:

            data =[ pci_l, 0x23, address_size_length]
            self.__add_to_list(data, memory_address)
            self.__add_to_list(data, memory_size)
        else:
            data =[ pci_l + len(response), 0x63, address_size_length]
            self.__add_to_list(data, memory_address)
            self.__add_to_list(data, memory_size)
            data = data + response

        self.send_frame(id, data)

    def request_transfer_exit(self, id, response = False):
        if response:
            data = [2, 0x37, 0]
        else:
             data = [2, 0x77, 0]

        self.send_frame(id, data)

    def clear_diagnostic_information(self, id, group_of_dtc, response=False):
        pci_l = len(group_of_dtc) + 1
        if response:
            data = [pci_l, 0x14] + group_of_dtc
        else:
            data = [1, 0x54]
        
        self.send_frame(id, data)

    def negative_response(self, id, sid, nrc): 
        data = [3, 0x7F, sid, nrc]
      
    
        self.send_frame(id, data)

    #ruben
    def routine_control(self, id, sub_funct, routine_id, response = False):
        if response:
            data = [4, 0x71, sub_funct, routine_id // 0x100, routine_id % 0x100]
        else:
            data = [4, 0x31, sub_funct, routine_id // 0x100, routine_id % 0x100]

        self.send_frame(id, data)
        
    def authntication_seed(self, id, seed = []):
        length_seed = len(seed)
        if length_seed > 0:
            data = [length_seed + 2, 0x69, 0x1] + seed
        else:
            data = [2, 0x29, 0x1]
            

        self.send_frame(id, data)

    def authntication_key(self, id, key = []):
        length_key = len(key)
        if length_key > 0:
            data = [length_key + 2, 0x29, 0x2] + key
        else:
            data = [2, 0x69, 0x2]

        self.send_frame(id, data)
    
    def tester_present(self, id, response = False):
        if response:
            data = [2, 0x7E, 0]
        else:
            data = [2, 0x3E, 0]

        self.send_frame(id, data)
    
    def access_timing_parameters(self, id, sub_function, response = False):
        if response:
            data = [2, 0x83, sub_function]
        else:
            data = [2, 0xC3, sub_function]

        self.send_frame(id, data)

    def request_download(self, id, data_format_identifier, memory_address, memory_size):
        address_length = (self.__count_digits(memory_address)+1)//2
        size_length = (self.__count_digits(memory_size)+1)//2
        memory_length = size_length * 0x10 + address_length
        data = [size_length+ address_length + 3,0x34, data_format_identifier, memory_length]
        self.__add_to_list(data, memory_address)
        self.__add_to_list(data, memory_size)

        self.send_frame(id, data)

    def request_download_response(self, id, max_number_block):
        mnb_length = (self.__count_digits(max_number_block)+1)//2
        data = [mnb_length + 2, 0x74, mnb_length * 0x10]
        self.__add_to_list(data, max_number_block)

        self.send_frame(id, data)

    def transfer_data(self, id, block_sequence_counter, transfer_data = {}):
        if len(transfer_data):
            if len(transfer_data) <=5:
                data = [len(transfer_data) + 2, 0x36, block_sequence_counter] + transfer_data
            else:
                print("ERROR: To many data to transfer, consider using Transfer_data_long!")
                return
        else:
            data =[0x2, 0x76, block_sequence_counter]
        self.send_frame(id, data)

    def transfer_data_long (self, id, block_sequence_counter, transfer_data, first_frame = True):
        if first_frame:
            data = [0x10,len(transfer_data) + 2,0x36,block_sequence_counter] + transfer_data[:4]
            self.send_frame(id, data)
        else:
            #Delete first 3 byts of data
            last_data = transfer_data[4:]
            for i in range(0, len(last_data)//7):
                data = [0x21 + i ] + last_data[i*7:(i+1)*7]
                self.send_frame(id, data)
            #Send remaining data
            if len(last_data)%7:
                data = [0x21 + len(last_data)//7] + last_data[len(last_data)- len(last_data)%7:]
                self.send_frame(id, data)

    def write_data_by_identifier(self, id, identifier, data_parameter):
        if len(data_parameter) > 0:
            if len(data_parameter) <=4:
                data =[len(data_parameter) +3 , 0x2E, identifier // 0x100, identifier%0x100] + data_parameter
            else:
                print("ERROR: To many data parameters, consider using write_data_by_identifier_long!")
                return
        else:
            data =[3, 0x6E, identifier // 0x100, identifier%0x100]
        
        self.send_frame(id,data)
            
    def write_data_by_identifier_long(self, id, identifier, data, first_frame=True):
        self.__generate_long_response(id, 0x2E, identifier, data, first_frame)


    def __generate_long_response(self, id, sid, identifier, response , first_frame):
        if first_frame:
            data = [0x10,len(response) + 3, sid, identifier//0x100, identifier%0x100] + response[:3]
            self.send_frame(id, data)
        else:
            last_data = response[3:]
            for i in range(0, len(last_data)//7):
                data = [(0x21 + i%8)] + last_data[i*7:(i+1)*7]
                self.send_frame(id, data)
            if len(last_data)%7:
                data = [0x21 + len(last_data)//7] + last_data[len(last_data)- len(last_data)%7:]
                self.send_frame(id, data) 

    def __add_to_list(self, data_list, number):
        temp_list = []
        while number:
            temp_list.append(number%0x100)
            number //=0x100
        data_list.extend(temp_list[::-1])


    def __count_digits(self, number):
        digits = 0
        while number:
            digits += 1
            number //=10
        return digits


can_interface = "vcan0"
id = 0x1234
data = [0,1,2,3,4,5]

generateFrame = GenerateFrame(can_interface)

generateFrame.read_data_by_identifier(id, 0x1234, data)
generateFrame.read_data_by_identifier_long(id, 0x1234, data)
generateFrame.read_data_by_identifier_long(id, 0x1234, data, False)

generateFrame.bus.shutdown()

